README
=========================================================================
| README for CCSS                                                       |
|                                                                       |
| http://ccss.forge.ocamlcore.org/                                      |
|                                                                       |
| Dario Teixeira <dario.teixeira@nleyten.com>                           |
=========================================================================


1. Overview
===========

CCSS is a preprocessor/pretty-printer for CSS (Cascading Style Sheets).
It extends the CSS language with support for declaration of variables and
basic arithmetic operations (addition, subtraction, multiplication, division).
The programme is supposed to be used as a filter: it reads the CSS source
from stdin and outputs its result on stdout.


2. Features
===========

2.1. Variables for expressions and mixins
=========================================

With CCSS, you may declare and use variables in your CSS code (in fact, CCSS
variables are actually constants, since they are assigned upon declaration and
may not be changed).  Variable identifiers must begin with an uppercase letter,
and be followed by any combination of letters, numbers, and the characters
'-' (dash) and '_' (underscore).

There are two use cases for variables.  The first consists in assigning any
CSS expression (not only quantities) to a variable.  The assigned expression
can subsequently be referenced by the variable identifier.  The code below
demonstrates this situation:

Foo: 20em;
Bar: 1px solid black;

h1  {
    width: Foo;
    border: Bar;
    }


The second use of variables is to declare mixins, ie, declaration blocks
that can be included within subsequent declaration blocks.  The code below
illustrates the use of mixins:

Inverse_scheme:
    {
    color: #fff;
    background: #000;
    }

h1
    {
    Inverse_scheme;
    font-weight: bold;
    }


2.2. Arithmetic
===============

CCSS extends CSS expressions with basic arithmetic operations (addition,
subtraction, multiplication, and division).  The operands must be CSS
quantities (either dimensionless or with an attached unit), or other
expressions that ultimately resolve into a quantity.  Moreover, variables
whose value is a quantity (or an expression which resolves into a quantity)
may also be used as operands.

The operators are '+', '-', '*', and '÷'.  Note that multiplication and
division have precedence over addition and subtraction, but you may use
parentheses to group operations.  Consider thus the following input:

Left: 10em;
Right: 5em;
Total: Left + Right;

h1  {
    padding: (1.5em + 0.5em) * 2;
    width: 2 * Total;
    }


CCSS will produce the following output:

h1
    {
    padding: 4em;
    width: 30em;
    }


The reader will have noticed that CCSS must be unit-aware when performing
arithmetic.  As a matter of fact, the programme performs a basic sanity
check of units, and will complain if you try, for example, to add "1em" with
"10px".  By default, CCSS will not make any attempt to convert units even if
they are convertible, such "cm" and "mm".  If you wish for CCSS to attempt
unit conversion, please provide option "--convert" on the command line
(short version "-c").

Units can be grouped into four categories, and conversion is possible if the
units belong to the same category.  Upon conversion, the result will be the
same unit as the first operand.  The categories and corresponding units are
as follows:

     length: mm, cm, in, pt, pc
      angle: deg, grad, rad
       time: ms, s
  frequency: hz, khz

As an illustration of unit conversion, the result for all the following
arithmetic operations is the same, "2in":

h1
    {
    foo1: 1in + 1in;
    foo2: 1in + 2.54cm;
    foo3: 1in + 25.4mm;
    foo4: 1in + 72pt;
    foo5: 1in + 6pc;
    }


3. Dependencies
===============

Menhir is used as the parser generator [1], while scanning is done with Sedlex [2].
Other requirements are Batteries [3] and OCaml-re [4].


4. Building and installing
==========================

The build system is generated with OASIS.  Run 'configure' if you wish to modify
the default install destination. Enter 'make' and 'make install' to build and
install the binary executable.


5. License
==========

CCSS is distributed under the terms of the GNU GPL version 2.
See LICENSE file for full license text.


References
==========

[1] http://cristal.inria.fr/~fpottier/menhir/
[2] https://github.com/alainfrisch/sedlex
[3] http://batteries.forge.ocamlcore.org/
[4] https://github.com/ocaml/ocaml-re



CHANGELOG
-------------------------
(2016-01-25) version 1.6:

  * Better support for complex media queries.
  * Add unit tests based on Alcotest.
  * Migrate from OCaml-pcre to OCaml-re.
  * Migrate from Ulex to Sedlex.
  * Drop '_t' suffix from types internally.

-------------------------
(2014-05-21) version 1.5:

  * Support for @keyframes (CSS animations).
  * Minor grammar simplification: @charset now lumped with other at-rules.

-------------------------
(2013-09-27) version 1.4:

  * Support for mixins.
  * Support for Unicode ranges.

-------------------------
(2012-09-28) version 1.3:

  * Now using OASIS for build system.
  * Migrated from ExtLib to Batteries.

-------------------------
(2010-03-24) version 1.2:

  * Handling of !important declarations.
  * More leniency in characters allowed in identifiers.
  * Allow division of quantities with the same unit (result is scalar).
  * Always allow addition/subtraction when one of the operands is the scalar 0.

-------------------------
(2010-03-12) version 1.1:

  * Added support for unit conversion.

-------------------------
(2010-03-10) version 1.0:

  * First public release.



tests/test_ccss.ml
(********************************************************************************)
(*  Test_ccss.ml
    Copyright (c) 2010-2016 Dario Teixeira <dario.teixeira@nleyten.com>
    This software is distributed under the terms of the GNU GPL version 2.
    See LICENSE file for full license text.
*)
(********************************************************************************)

module String = BatString


(********************************************************************************)
(** {1 Functions and values}                                                    *)
(********************************************************************************)

let execute source () =
    let ch = open_in source in
    let source_contents = BatPervasives.input_all ch in
    close_in ch;
    let (in_ch, out_ch) = Unix.open_process "ccss --convert" in
    output_string out_ch source_contents;
    flush out_ch;
    close_out out_ch;
    let target_contents = BatPervasives.input_all in_ch in
    let _ = Unix.close_process (in_ch, out_ch) in
    let correct = String.slice ~last:(-4) source ^ "css" in
    let ch = open_in correct in
    let correct_contents = BatPervasives.input_all ch in
    close_in ch;
    Alcotest.(check string) source correct_contents target_contents

let build_test () =
    let process filename =
        let ch = open_in filename in
        let line = input_line ch in
        close_in ch;
        (filename, String.slice ~first:3 line, `Quick, execute filename) in
    let hnd = Unix.opendir "." in
    let rec loop accum = match Unix.readdir hnd with
        | filename ->
            if String.ends_with filename ".ccss"
            then loop (process filename :: accum)
            else loop accum
        | exception End_of_file ->
            Unix.closedir hnd;
            accum in
    loop [] |> List.sort Pervasives.compare |> List.map (fun (_, d, q, f) -> (d, q, f))

let tests =
    [
    ("CCSS testset", build_test ());
    ]

let () =
    Alcotest.run "CCSS tests" tests



src/ast.ml
(********************************************************************************)
(*  Ast.ml
    Copyright (c) 2010-2016 Dario Teixeira <dario.teixeira@nleyten.com>
    This software is distributed under the terms of the GNU GPL version 2.
    See LICENSE file for full license text.
*)
(********************************************************************************)

(** Defines the AST for parsed CSS stylesheets.
*)

(********************************************************************************)
(** {1 Type definitions}                                                        *)
(********************************************************************************)

(********************************************************************************)
(** {2 Top-level statements}                                                    *)
(********************************************************************************)

type t = statement list

and statement =
    [ `Atrule of prefix option * atrule
    | `Rule of rule
    | `Vardecl of variable * vardecl
    ]

and atrule =
    [ `Charset of string
    | `Import of source * medium list option
    | `Media of medium list * rule list
    | `Page of pseudo_page option * declaration list
    | `Fontface of declaration list
    | `Keyframes of string * keyframe_block list
    ]

and prefix = string

and source =
    [ `String of string
    | `Uri of string
    ]

and medium = string

and pseudo_page = string

and keyframe_block = keyframe_sel * declaration list

and keyframe_sel =
    [ `Ident of string
    | `Calc of calc
    ]

and rule = selector list * declaration list

and variable = Lexing.position * string

and vardecl =
    [ `Expr of expression
    | `Mixin of declaration list
    ]


(********************************************************************************)
(** {2 Selectors}                                                               *)
(********************************************************************************)

and selector = simplesel * (combinator * simplesel) list

and simplesel =
    [ `Explicit of element * qualifier list
    | `Generic of qualifier * qualifier list
    ]

and combinator =
    [ `Descendant
    | `General_sibling
    | `Adjacent_sibling
    | `Child
    ]

and element =
    [ `Tag of string
    | `Universal
    ]

and qualifier =
    [ `Id of string
    | `Class of string
    | `Attr of string * attr
    | `Pseudo_class of string
    | `Pseudo_element of string
    | `Sel_func of string * func
    ]

and func =
    [ `Qualified of qualifier list
    | `Nth of string
    ]

and attr =
    [ `Attr_exists
    | `Attr_equals of string
    | `Attr_includes of string
    | `Attr_dashmatch of string
    | `Attr_prefix of string
    | `Attr_suffix of string
    | `Attr_substring of string
    ]


(********************************************************************************)
(** {2 Declarations}                                                            *)
(********************************************************************************)

and declaration =
    [ `Property of property * expression * important
    | `Varref of variable
    ]

and property = string

and important = bool

and expression = sentence list

and sentence = term list

and term =
    [ `Calc of calc
    | `String of string
    | `Ident of string
    | `Uri of string
    | `Hash of string
    | `Urange of string
    | `Term_func of string * expression
    | `Slash
    ]

and calc =
    [ `Varref of variable
    | `Quantity of quantity
    | `Sum of Lexing.position * calc * calc
    | `Sub of Lexing.position * calc * calc
    | `Mul of Lexing.position * calc * calc
    | `Div of Lexing.position * calc * calc
    ]

and quantity = float * string option



src/ccss.ml
(********************************************************************************)
(*  Ccss.ml
    Copyright (c) 2010-2016 Dario Teixeira <dario.teixeira@nleyten.com>
    This software is distributed under the terms of the GNU GPL version 2.
    See LICENSE file for full license text.
*)
(********************************************************************************)

open Lexing
open Parser


(********************************************************************************)
(** {1 Exceptions}                                                              *)
(********************************************************************************)

exception Scanning_error of Lexing.position * string
exception Syntax_error of Lexing.position


(********************************************************************************)
(** {1 Functions and values}                                                    *)
(********************************************************************************)

let menhir_with_sedlex menhir_parser lexbuf =
    let position = ref
        {
        pos_fname = "";
        pos_lnum = 1;
        pos_bol = 0;
        pos_cnum = 0;
        } in
    let lexer_maker () =
        let ante_position = !position in
        let (nlines, token) = Scanner.main_scanner 0 lexbuf in
        let () = position := {!position with pos_lnum = !position.pos_lnum + nlines;} in
        let post_position = !position
        in (token, ante_position, post_position) in
    let revised_parser = MenhirLib.Convert.Simplified.traditional2revised menhir_parser
    in try
        revised_parser lexer_maker
    with
        | Scanner.Error x -> raise (Scanning_error (!position, x))
        | Parser.Error    -> raise (Syntax_error !position)


let string_of_exc convert = function
    | Scanning_error (pos, x) ->
        Printf.sprintf "Scanning error on line %d: cannot interpret '%s'.\n" pos.pos_lnum x
    | Syntax_error pos ->
        Printf.sprintf "Syntax error on line %d.\n" pos.pos_lnum
    | Printer.Variable_redeclared (pos, id) ->
        Printf.sprintf "Attempt to redefine variable '%s' in line %d.\n" id pos.pos_lnum
    | Printer.Variable_undeclared (pos, id) ->
        Printf.sprintf "Variable '%s' referenced in line %d has not been declared.\n" id pos.pos_lnum
    | Printer.Expected_mixin_over_expression (pos, id) ->
        Printf.sprintf "In line %d, variable '%s' refers to an expression, but a mixin was expected in this context.\n" pos.pos_lnum id
    | Printer.Expected_expression_over_mixin (pos, id) ->
        Printf.sprintf "In line %d, variable '%s' refers to a mixin, but an expression was expected in this context.\n" pos.pos_lnum id
    | Printer.Invalid_arithmetic (pos, op) ->
        Printf.sprintf "Invalid arithmetic in line %d: attempt to %s with non-numeric expression.\n" pos.pos_lnum op
    | Printer.Invalid_units (pos, op, u1, u2) ->
        let error = Printf.sprintf "Invalid use of units in line %d: attempt to %s %s and %s.\n" pos.pos_lnum op u1 u2 in
        if convert
        then error
        else Printf.sprintf "%s\nHint: the '--convert' option may be used to attempt unit conversion, where applicable.\n" error
    | exc ->
        Printexc.to_string exc


let () =
    let convert = Options.parse ()
    in try
        let lexbuf = Sedlexing.Utf8.from_channel stdin in
        let css = menhir_with_sedlex Parser.stylesheet lexbuf in
        let out = Printer.sprint convert css
        in print_string out
    with exc ->
        let msg = string_of_exc convert exc in
        output_string stderr msg;
        exit 1



src/options.ml
(********************************************************************************)
(*  Options.ml
    Copyright (c) 2010-2016 Dario Teixeira <dario.teixeira@nleyten.com>
    This software is distributed under the terms of the GNU GPL version 2.
    See LICENSE file for full license text.
*)
(********************************************************************************)

open BatOptParse


(********************************************************************************)
(** {1 Functions and values}                                                    *)
(********************************************************************************)

let options = OptParser.make ()

let convert_opt = StdOpt.store_true ()


let () =
    let general = OptParser.add_group options "General options"
    in OptParser.add options ~group:general ~short_name:'c' ~long_name:"convert" ~help:"Attempt unit conversion" convert_opt


let parse () = match OptParser.parse_argv options with
    | hd::tl ->
        OptParser.usage options ();
        OptParser.error options "Error: invalid usage";
        raise Exit
    | [] ->
        Opt.get convert_opt



myocamlbuild.ml
(* OASIS_START *)
(* DO NOT EDIT (digest: e47f0962d813dd33c9fc90cad2c6845c) *)
module OASISGettext = struct
(* # 21 "/home/dario/.godi312/build/apps/apps-oasis/work/oasis-0.3.0~rc6/src/oasis/OASISGettext.ml" *)

  let ns_ str =
    str

  let s_ str =
    str

  let f_ (str : ('a, 'b, 'c, 'd) format4) =
    str

  let fn_ fmt1 fmt2 n =
    if n = 1 then
      fmt1^^""
    else
      fmt2^^""

  let init =
    []

end

module OASISExpr = struct
(* # 21 "/home/dario/.godi312/build/apps/apps-oasis/work/oasis-0.3.0~rc6/src/oasis/OASISExpr.ml" *)



  open OASISGettext

  type test = string 

  type flag = string 

  type t =
    | EBool of bool
    | ENot of t
    | EAnd of t * t
    | EOr of t * t
    | EFlag of flag
    | ETest of test * string
    

  type 'a choices = (t * 'a) list 

  let eval var_get t =
    let rec eval' =
      function
        | EBool b ->
            b

        | ENot e ->
            not (eval' e)

        | EAnd (e1, e2) ->
            (eval' e1) && (eval' e2)

        | EOr (e1, e2) ->
            (eval' e1) || (eval' e2)

        | EFlag nm ->
            let v =
              var_get nm
            in
              assert(v = "true" || v = "false");
              (v = "true")

        | ETest (nm, vl) ->
            let v =
              var_get nm
            in
              (v = vl)
    in
      eval' t

  let choose ?printer ?name var_get lst =
    let rec choose_aux =
      function
        | (cond, vl) :: tl ->
            if eval var_get cond then
              vl
            else
              choose_aux tl
        | [] ->
            let str_lst =
              if lst = [] then
                s_ "<empty>"
              else
                String.concat
                  (s_ ", ")
                  (List.map
                     (fun (cond, vl) ->
                        match printer with
                          | Some p -> p vl
                          | None -> s_ "<no printer>")
                     lst)
            in
              match name with
                | Some nm ->
                    failwith
                      (Printf.sprintf
                         (f_ "No result for the choice list '%s': %s")
                         nm str_lst)
                | None ->
                    failwith
                      (Printf.sprintf
                         (f_ "No result for a choice list: %s")
                         str_lst)
    in
      choose_aux (List.rev lst)

end


# 117 "myocamlbuild.ml"
module BaseEnvLight = struct
(* # 21 "/home/dario/.godi312/build/apps/apps-oasis/work/oasis-0.3.0~rc6/src/base/BaseEnvLight.ml" *)

  module MapString = Map.Make(String)

  type t = string MapString.t

  let default_filename =
    Filename.concat
      (Sys.getcwd ())
      "setup.data"

  let load ?(allow_empty=false) ?(filename=default_filename) () =
    if Sys.file_exists filename then
      begin
        let chn =
          open_in_bin filename
        in
        let st =
          Stream.of_channel chn
        in
        let line =
          ref 1
        in
        let st_line =
          Stream.from
            (fun _ ->
               try
                 match Stream.next st with
                   | '\n' -> incr line; Some '\n'
                   | c -> Some c
               with Stream.Failure -> None)
        in
        let lexer =
          Genlex.make_lexer ["="] st_line
        in
        let rec read_file mp =
          match Stream.npeek 3 lexer with
            | [Genlex.Ident nm; Genlex.Kwd "="; Genlex.String value] ->
                Stream.junk lexer;
                Stream.junk lexer;
                Stream.junk lexer;
                read_file (MapString.add nm value mp)
            | [] ->
                mp
            | _ ->
                failwith
                  (Printf.sprintf
                     "Malformed data file '%s' line %d"
                     filename !line)
        in
        let mp =
          read_file MapString.empty
        in
          close_in chn;
          mp
      end
    else if allow_empty then
      begin
        MapString.empty
      end
    else
      begin
        failwith
          (Printf.sprintf
             "Unable to load environment, the file '%s' doesn't exist."
             filename)
      end

  let var_get name env =
    let rec var_expand str =
      let buff =
        Buffer.create ((String.length str) * 2)
      in
        Buffer.add_substitute
          buff
          (fun var ->
             try
               var_expand (MapString.find var env)
             with Not_found ->
               failwith
                 (Printf.sprintf
                    "No variable %s defined when trying to expand %S."
                    var
                    str))
          str;
        Buffer.contents buff
    in
      var_expand (MapString.find name env)

  let var_choose lst env =
    OASISExpr.choose
      (fun nm -> var_get nm env)
      lst
end


# 215 "myocamlbuild.ml"
module MyOCamlbuildFindlib = struct
(* # 21 "/home/dario/.godi312/build/apps/apps-oasis/work/oasis-0.3.0~rc6/src/plugins/ocamlbuild/MyOCamlbuildFindlib.ml" *)

  (** OCamlbuild extension, copied from 
    * http://brion.inria.fr/gallium/index.php/Using_ocamlfind_with_ocamlbuild
    * by N. Pouillard and others
    *
    * Updated on 2009/02/28
    *
    * Modified by Sylvain Le Gall 
    *)
  open Ocamlbuild_plugin

  (* these functions are not really officially exported *)
  let run_and_read = 
    Ocamlbuild_pack.My_unix.run_and_read

  let blank_sep_strings = 
    Ocamlbuild_pack.Lexers.blank_sep_strings

  let split s ch =
    let x = 
      ref [] 
    in
    let rec go s =
      let pos = 
        String.index s ch 
      in
        x := (String.before s pos)::!x;
        go (String.after s (pos + 1))
    in
      try
        go s
      with Not_found -> !x

  let split_nl s = split s '\n'

  let before_space s =
    try
      String.before s (String.index s ' ')
    with Not_found -> s

  (* this lists all supported packages *)
  let find_packages () =
    List.map before_space (split_nl & run_and_read "ocamlfind list")

  (* this is supposed to list available syntaxes, but I don't know how to do it. *)
  let find_syntaxes () = ["camlp4o"; "camlp4r"]

  (* ocamlfind command *)
  let ocamlfind x = S[A"ocamlfind"; x]

  let dispatch =
    function
      | Before_options ->
          (* by using Before_options one let command line options have an higher priority *)
          (* on the contrary using After_options will guarantee to have the higher priority *)
          (* override default commands by ocamlfind ones *)
          Options.ocamlc     := ocamlfind & A"ocamlc";
          Options.ocamlopt   := ocamlfind & A"ocamlopt";
          Options.ocamldep   := ocamlfind & A"ocamldep";
          Options.ocamldoc   := ocamlfind & A"ocamldoc";
          Options.ocamlmktop := ocamlfind & A"ocamlmktop"
                                  
      | After_rules ->
          
          (* When one link an OCaml library/binary/package, one should use -linkpkg *)
          flag ["ocaml"; "link"; "program"] & A"-linkpkg";
          
          (* For each ocamlfind package one inject the -package option when
           * compiling, computing dependencies, generating documentation and
           * linking. *)
          List.iter 
            begin fun pkg ->
              flag ["ocaml"; "compile";  "pkg_"^pkg] & S[A"-package"; A pkg];
              flag ["ocaml"; "ocamldep"; "pkg_"^pkg] & S[A"-package"; A pkg];
              flag ["ocaml"; "doc";      "pkg_"^pkg] & S[A"-package"; A pkg];
              flag ["ocaml"; "link";     "pkg_"^pkg] & S[A"-package"; A pkg];
              flag ["ocaml"; "infer_interface"; "pkg_"^pkg] & S[A"-package"; A pkg];
            end 
            (find_packages ());

          (* Like -package but for extensions syntax. Morover -syntax is useless
           * when linking. *)
          List.iter begin fun syntax ->
          flag ["ocaml"; "compile";  "syntax_"^syntax] & S[A"-syntax"; A syntax];
          flag ["ocaml"; "ocamldep"; "syntax_"^syntax] & S[A"-syntax"; A syntax];
          flag ["ocaml"; "doc";      "syntax_"^syntax] & S[A"-syntax"; A syntax];
          flag ["ocaml"; "infer_interface"; "syntax_"^syntax] & S[A"-syntax"; A syntax];
          end (find_syntaxes ());

          (* The default "thread" tag is not compatible with ocamlfind.
           * Indeed, the default rules add the "threads.cma" or "threads.cmxa"
           * options when using this tag. When using the "-linkpkg" option with
           * ocamlfind, this module will then be added twice on the command line.
           *                        
           * To solve this, one approach is to add the "-thread" option when using
           * the "threads" package using the previous plugin.
           *)
          flag ["ocaml"; "pkg_threads"; "compile"] (S[A "-thread"]);
          flag ["ocaml"; "pkg_threads"; "doc"] (S[A "-I"; A "+threads"]);
          flag ["ocaml"; "pkg_threads"; "link"] (S[A "-thread"]);
          flag ["ocaml"; "pkg_threads"; "infer_interface"] (S[A "-thread"])

      | _ -> 
          ()

end

module MyOCamlbuildBase = struct
(* # 21 "/home/dario/.godi312/build/apps/apps-oasis/work/oasis-0.3.0~rc6/src/plugins/ocamlbuild/MyOCamlbuildBase.ml" *)

  (** Base functions for writing myocamlbuild.ml
      @author Sylvain Le Gall
    *)



  open Ocamlbuild_plugin
  module OC = Ocamlbuild_pack.Ocaml_compiler

  type dir = string 
  type file = string 
  type name = string 
  type tag = string 

(* # 56 "/home/dario/.godi312/build/apps/apps-oasis/work/oasis-0.3.0~rc6/src/plugins/ocamlbuild/MyOCamlbuildBase.ml" *)

  type t =
      {
        lib_ocaml: (name * dir list) list;
        lib_c:     (name * dir * file list) list; 
        flags:     (tag list * (spec OASISExpr.choices)) list;
        (* Replace the 'dir: include' from _tags by a precise interdepends in
         * directory.
         *)
        includes:  (dir * dir list) list; 
      } 

  let env_filename =
    Pathname.basename 
      BaseEnvLight.default_filename

  let dispatch_combine lst =
    fun e ->
      List.iter 
        (fun dispatch -> dispatch e)
        lst 

  let tag_libstubs nm =
    "use_lib"^nm^"_stubs"

  let nm_libstubs nm =
    nm^"_stubs"

  let dispatch t e = 
    let env = 
      BaseEnvLight.load 
        ~filename:env_filename 
        ~allow_empty:true
        ()
    in
      match e with 
        | Before_options ->
            let no_trailing_dot s =
              if String.length s >= 1 && s.[0] = '.' then
                String.sub s 1 ((String.length s) - 1)
              else
                s
            in
              List.iter
                (fun (opt, var) ->
                   try 
                     opt := no_trailing_dot (BaseEnvLight.var_get var env)
                   with Not_found ->
                     Printf.eprintf "W: Cannot get variable %s" var)
                [
                  Options.ext_obj, "ext_obj";
                  Options.ext_lib, "ext_lib";
                  Options.ext_dll, "ext_dll";
                ]

        | Before_rules ->
          (* TODO: move this into its own file and conditionnaly include it, if
           * needed.
           *)
          (* OCaml cmxs rules: cmxs available in ocamlopt but not ocamlbuild.
             Copied from ocaml_specific.ml in ocamlbuild sources. *)
          let has_native_dynlink =
            try
              bool_of_string (BaseEnvLight.var_get "native_dynlink" env)
            with Not_found ->
              false
          in
          if has_native_dynlink && String.sub Sys.ocaml_version 0 4 = "3.11" then
            begin
              let ext_lib = !Options.ext_lib in
              let ext_obj = !Options.ext_obj in
              let ext_dll = !Options.ext_dll in
              let x_o = "%"-.-ext_obj in
              let x_a = "%"-.-ext_lib in
              let x_dll = "%"-.-ext_dll in
              let x_p_o = "%.p"-.-ext_obj in
              let x_p_a = "%.p"-.-ext_lib in
              let x_p_dll = "%.p"-.-ext_dll in

              rule "ocaml: mldylib & p.cmx* & p.o* -> p.cmxs & p.so"
                   ~tags:["ocaml"; "native"; "profile"; "shared"; "library"]
                   ~prods:["%.p.cmxs"; x_p_dll]
                   ~dep:"%.mldylib"
                   (OC.native_profile_shared_library_link_mldylib
                      "%.mldylib" "%.p.cmxs");

              rule "ocaml: mldylib & cmx* & o* -> cmxs & so"
                   ~tags:["ocaml"; "native"; "shared"; "library"]
                   ~prods:["%.cmxs"; x_dll]
                   ~dep:"%.mldylib"
                   (OC.native_shared_library_link_mldylib
                      "%.mldylib" "%.cmxs");

              rule "ocaml: p.cmx & p.o -> p.cmxs & p.so"
                   ~tags:["ocaml"; "native"; "profile"; "shared"; "library"]
                   ~prods:["%.p.cmxs"; x_p_dll]
                   ~deps:["%.p.cmx"; x_p_o]
                   (OC.native_shared_library_link ~tags:["profile"]
                                                  "%.p.cmx" "%.p.cmxs");

              rule "ocaml: p.cmxa & p.a -> p.cmxs & p.so"
                   ~tags:["ocaml"; "native"; "profile"; "shared"; "library"]
                   ~prods:["%.p.cmxs"; x_p_dll]
                   ~deps:["%.p.cmxa"; x_p_a]
                   (OC.native_shared_library_link ~tags:["profile"; "linkall"]
                                                  "%.p.cmxa" "%.p.cmxs");

              rule "ocaml: cmx & o -> cmxs"
                   ~tags:["ocaml"; "native"; "shared"; "library"]
                   ~prods:["%.cmxs"]
                   ~deps:["%.cmx"; x_o]
                   (OC.native_shared_library_link "%.cmx" "%.cmxs");

              rule "ocaml: cmx & o -> cmxs & so"
                   ~tags:["ocaml"; "native"; "shared"; "library"]
                   ~prods:["%.cmxs"; x_dll]
                   ~deps:["%.cmx"; x_o]
                   (OC.native_shared_library_link "%.cmx" "%.cmxs");

              rule "ocaml: cmxa & a -> cmxs & so"
                   ~tags:["ocaml"; "native"; "shared"; "library"]
                   ~prods:["%.cmxs"; x_dll]
                   ~deps:["%.cmxa"; x_a]
                   (OC.native_shared_library_link ~tags:["linkall"]
                                                  "%.cmxa" "%.cmxs");
            end

        | After_rules -> 
            (* Declare OCaml libraries *)
            List.iter 
              (function
                 | nm, [] ->
                     ocaml_lib nm
                 | nm, dir :: tl ->
                     ocaml_lib ~dir:dir (dir^"/"^nm);
                     List.iter 
                       (fun dir -> 
                          List.iter
                            (fun str ->
                               flag ["ocaml"; "use_"^nm; str] (S[A"-I"; P dir]))
                            ["compile"; "infer_interface"; "doc"])
                       tl)
              t.lib_ocaml;

            (* Declare directories dependencies, replace "include" in _tags. *)
            List.iter 
              (fun (dir, include_dirs) ->
                 Pathname.define_context dir include_dirs)
              t.includes;

            (* Declare C libraries *)
            List.iter
              (fun (lib, dir, headers) ->
                   (* Handle C part of library *)
                   flag ["link"; "library"; "ocaml"; "byte"; tag_libstubs lib]
                     (S[A"-dllib"; A("-l"^(nm_libstubs lib)); A"-cclib";
                        A("-l"^(nm_libstubs lib))]);

                   flag ["link"; "library"; "ocaml"; "native"; tag_libstubs lib]
                     (S[A"-cclib"; A("-l"^(nm_libstubs lib))]);
                        
                   flag ["link"; "program"; "ocaml"; "byte"; tag_libstubs lib]
                     (S[A"-dllib"; A("dll"^(nm_libstubs lib))]);

                   (* When ocaml link something that use the C library, then one
                      need that file to be up to date.
                    *)
                   dep  ["link"; "ocaml"; "program"; tag_libstubs lib]
                     [dir/"lib"^(nm_libstubs lib)^"."^(!Options.ext_lib)];

                   dep  ["compile"; "ocaml"; "program"; tag_libstubs lib]
                     [dir/"lib"^(nm_libstubs lib)^"."^(!Options.ext_lib)];

                   (* TODO: be more specific about what depends on headers *)
                   (* Depends on .h files *)
                   dep ["compile"; "c"] 
                     headers;

                   (* Setup search path for lib *)
                   flag ["link"; "ocaml"; "use_"^lib] 
                     (S[A"-I"; P(dir)]);
              )
              t.lib_c;

              (* Add flags *)
              List.iter
              (fun (tags, cond_specs) ->
                 let spec = 
                   BaseEnvLight.var_choose cond_specs env
                 in
                   flag tags & spec)
              t.flags
        | _ -> 
            ()

  let dispatch_default t =
    dispatch_combine 
      [
        dispatch t;
        MyOCamlbuildFindlib.dispatch;
      ]

end


# 548 "myocamlbuild.ml"
open Ocamlbuild_plugin;;
let package_default =
  {MyOCamlbuildBase.lib_ocaml = []; lib_c = []; flags = []; includes = []; }
  ;;

let dispatch_default = MyOCamlbuildBase.dispatch_default package_default;;

# 557 "myocamlbuild.ml"
(* OASIS_STOP *)
Ocamlbuild_plugin.dispatch dispatch_default;;
